File: scratch/test_filtering.py
import re

# Load portfolio.html
with open('portfolio.html', 'r', encoding='utf-8') as f:
    html = f.read()

# Setup simulation state
isMainPage = False
isEdit = False

# Represent buttons and cards
class Element:
    def __init__(self, tag_name, attrs, text=""):
        self.tag_name = tag_name
        self.attrs = attrs
        self.dataset = attrs
        self.text = text
        self.style = {}
        self.classes = attrs.get('class', '').split()
        
    def getAttribute(self, name):
        return self.attrs.get(name)

# Parse sub-filter buttons
sub_filters = []
btn_matches = re.finditer(r'<button class="sub-filter-btn([^"]*)"[^>]*data-subcat="([^"]*)"[^>]*>(.*?)</button>', html)
for m in btn_matches:
    class_str, subcat, text = m.groups()
    classes = "sub-filter-btn " + class_str
    sub_filters.append(Element('button', {'class': classes, 'data-subcat': subcat}, text))

# Parse cards
cards = []
# We split by '<div class="portfolio-card' to find all card elements
parts = html.split('<div class="portfolio-card')
for part in parts[1:]:
    attr_match = re.match(r'^([^>]+)>', part)
    if attr_match:
        attrs_str = attr_match.group(1)
        # Parse attributes
        attrs = {}
        for attr_m in re.finditer(r'([\w\-]+)="([^"]*)"', attrs_str):
            attrs[attr_m.group(1)] = attr_m.group(2)
        
        # Get title
        title_match = re.search(r'<h3[^>]*>(.*?)</h3>', part)
        title = title_match.group(1).strip() if title_match else ""
        
        cards.append(Element('div', attrs, title))

print(f"Parsed {len(sub_filters)} sub-filter buttons and {len(cards)} cards.")

def simulate_filter_click(active_cat, active_subcat):
    print(f"\n--- Simulating clicks: cat={active_cat}, subcat={active_subcat} ---")
    
    # Mock script.js filter logic
    cat = active_cat
    activeSubCat = active_subcat
    
    colsVal = 3 if cat == 'covers' else 2
    rowsVal = 0
    limit = colsVal * rowsVal
    activeCatFilteredCount = 0
    
    shown_cards = []
    hidden_cards = []
    
    for card in cards:
        cardCat = card.dataset.get('data-cat')
        card_subcat = card.dataset.get('data-subcat')
        
        shouldShow = False
        if cat == 'all' or cardCat == cat:
            if cat == 'covers' and activeSubCat != 'all' and card_subcat != activeSubCat:
                shouldShow = False
            else:
                if (cardCat in ['covers', 'formatting', 'paperback-covers']) and isMainPage and not isEdit:
                    shouldShow = False
                else:
                    if not isMainPage or isEdit:
                        shouldShow = True
        
        if shouldShow and not isMainPage and cat != 'all' and limit > 0:
            if activeCatFilteredCount >= limit:
                shouldShow = False
            else:
                activeCatFilteredCount += 1
                
        if shouldShow:
            card.style['display'] = 'block'
            shown_cards.append(card)
        else:
            card.style['display'] = 'none'
            hidden_cards.append(card)
            
    print(f"Total shown cards: {len(shown_cards)}")
    for idx, c in enumerate(shown_cards):
        print(f"  {idx+1}. Title: '{c.text}', Subcat: '{c.dataset.get('data-subcat')}', Cat: '{c.dataset.get('data-cat')}'")

# Simulate default load (Book Covers, All)
simulate_filter_click('covers', 'all')

# Simulate cookbooks
simulate_filter_click('covers', 'cookbooks-food-wine')

# Simulate health-fitness
simulate_filter_click('covers', 'health-fitness')

File: scratch/test_filtering.js
const fs = require('fs');
const { JSDOM } = require('jsdom');

const html = fs.readFileSync('portfolio.html', 'utf8');
const dom = new JSDOM(html, { runScripts: "outside-only" });
const document = dom.window.document;
const window = dom.window;

// Setup basic window mock values that script.js expects
window.location = { pathname: '/portfolio.html' };
const isMainPage = false;
const isEdit = false;

// Let's run the exact logic from script.js for category and subcategory filtering
// We will bind the events to the elements manually and test clicks.

// 1. Bind Main category buttons
const filterBtns = document.querySelectorAll('.filter-btn');
filterBtns.forEach(btn => {
    // Clone and replace to simulate the script.js behavior
    const oldBtn = btn.cloneNode(true);
    btn.parentNode.replaceChild(oldBtn, btn);
    
    oldBtn.addEventListener('click', function() {
        document.querySelectorAll('.filter-btn').forEach(b => b.classList.remove('active'));
        this.classList.add('active');
        const cat = this.dataset.cat;
        
        const subFilters = document.getElementById('book-covers-sub-filters');
        if (subFilters) {
            if (cat === 'covers') {
                subFilters.style.display = 'flex';
            } else {
                subFilters.style.display = 'none';
            }
        }
        
        const colsVal = parseInt(this.dataset.cols) || ( (cat === 'paperback-covers' || cat === 'formatting') ? 2 : 3 );
        const rowsVal = parseInt(this.dataset.rows) || 0;
        const limit = colsVal * rowsVal;
        let activeCatFilteredCount = 0;

        document.querySelectorAll('.portfolio-grid > .portfolio-card').forEach(card => {
            const cardCat = card.dataset.cat;
            let shouldShow = false;

            if (cat === 'all' || cardCat === cat) {
                const activeSubCatBtn = document.querySelector('.sub-filter-btn.active');
                const activeSubCat = activeSubCatBtn ? activeSubCatBtn.dataset.subcat : 'all';
                if (cat === 'covers' && activeSubCat !== 'all' && card.dataset.subcat !== activeSubCat) {
                    shouldShow = false;
                } else {
                    if ((cardCat === 'covers' || cardCat === 'formatting' || cardCat === 'paperback-covers') && isMainPage && !isEdit) {
                        shouldShow = false;
                    } else {
                        if (!isMainPage || isEdit) {
                            shouldShow = true;
                        }
                    }
                }
            }

            if (shouldShow && !isMainPage && cat !== 'all' && limit > 0) {
                if (activeCatFilteredCount >= limit) {
                    shouldShow = false;
                } else {
                    activeCatFilteredCount++;
                }
            }
            
            if (shouldShow) {
                card.style.display = 'block';
                card.classList.add('active'); // mock reveal observer
            } else {
                card.classList.remove('active');
                card.style.display = 'none';
            }
        });
    });
});

// 2. Bind Subcategory buttons
const subFilterBtns = document.querySelectorAll('.sub-filter-btn');
subFilterBtns.forEach(btn => {
    btn.addEventListener('click', function(e) {
        document.querySelectorAll('.sub-filter-btn').forEach(b => b.classList.remove('active'));
        this.classList.add('active');
        const activeMainFilter = document.querySelector('.filter-btn.active');
        if (activeMainFilter) activeMainFilter.click();
    });
});

// Let's run a click on default filter (Book Covers)
console.log("--- Initial Page Load (Book Covers filter clicked) ---");
const defaultFilter = document.querySelector('.filter-btn[data-cat="covers"]');
if (defaultFilter) {
    defaultFilter.click();
}

let visibleCovers = Array.from(document.querySelectorAll('.portfolio-grid > .portfolio-card')).filter(c => c.style.display === 'block');
console.log(`Visible covers after default click: ${visibleCovers.length}`);

// Let's click "Cookbooks, Food & Wine"
console.log("\n--- Clicking 'Cookbooks, Food & Wine' subcategory ---");
const cookbooksBtn = Array.from(document.querySelectorAll('.sub-filter-btn')).find(b => b.getAttribute('data-subcat') === 'cookbooks-food-wine');
if (cookbooksBtn) {
    cookbooksBtn.click();
} else {
    console.log("Cookbooks button NOT found!");
}

visibleCovers = Array.from(document.querySelectorAll('.portfolio-grid > .portfolio-card')).filter(c => c.style.display === 'block');
console.log(`Visible covers after cookbooks click: ${visibleCovers.length}`);
visibleCovers.forEach(c => {
    console.log(`  - Title: ${c.querySelector('h3').textContent.trim()}, Subcat: ${c.getAttribute('data-subcat')}, Img: ${c.querySelector('img').src}`);
});

// Let's click "Health & Fitness"
console.log("\n--- Clicking 'Health & Fitness' subcategory ---");
const healthBtn = Array.from(document.querySelectorAll('.sub-filter-btn')).find(b => b.getAttribute('data-subcat') === 'health-fitness');
if (healthBtn) {
    healthBtn.click();
} else {
    console.log("Health button NOT found!");
}

visibleCovers = Array.from(document.querySelectorAll('.portfolio-grid > .portfolio-card')).filter(c => c.style.display === 'block');
console.log(`Visible covers after health click: ${visibleCovers.length}`);
visibleCovers.forEach(c => {
    console.log(`  - Title: ${c.querySelector('h3').textContent.trim()}, Subcat: ${c.getAttribute('data-subcat')}, Img: ${c.querySelector('img').src}`);
});
